src/compiler/scala/tools/nsc/Reporting.scala

```diff
@@ -30,17 +30,28 @@ trait Reporting extends scala.reflect.internal.Reporting { self: ast.Positions w
       def this(what: String, booleanSetting: Settings#BooleanSetting) {
         this(what, () => booleanSetting, booleanSetting)
       }
-      val warnings = mutable.LinkedHashMap[Position, String]()
-      def warn(pos: Position, msg: String) =
+      val warnings = mutable.LinkedHashMap[Position, (String, String)]()
+      def warn(pos: Position, msg: String, since: String = "") =
         if (doReport()) reporter.warning(pos, msg)
-        else if (!(warnings contains pos)) warnings += ((pos, msg))
+        else if (!(warnings contains pos)) warnings += ((pos, (msg, since)))
       def summarize() =
         if (warnings.nonEmpty && (setting.isDefault || doReport())) {
-          val numWarnings  = warnings.size
-          val warningVerb  = if (numWarnings == 1) "was" else "were"
-          val warningCount = countElementsAsString(numWarnings, s"$what warning")
-
-          reporter.warning(NoPosition, s"there $warningVerb $warningCount; re-run with ${setting.name} for details")
+          val warningsGroupedBySince = warnings.groupBy{case (pos, (msg, since)) => since}
+          val deprecationSummary = warningsGroupedBySince.size > 1
+          warningsGroupedBySince.map{warnings => 
+            val numWarnings   = warnings._2.size
+            val warningsSince = if (warnings._1.nonEmpty) s" (since ${warnings._1})" else ""
+            val warningVerb   = if (numWarnings == 1) "was" else "were"
+            val warningCount  = countElementsAsString(numWarnings, s"$what warning")
+            val rerun         = if (deprecationSummary) "" else s"; re-run with ${setting.name} for details"
+            reporter.warning(NoPosition, s"there $warningVerb $warningCount$warningsSince$rerun")
+          }
+          if (deprecationSummary) {
+            val numWarnings   = warnings.size
+            val warningVerb   = if (numWarnings == 1) "was" else "were"
+            val warningCount  = countElementsAsString(numWarnings, s"$what warning")
+            reporter.warning(NoPosition, s"there $warningVerb $warningCount in total; re-run with ${setting.name} for details")
+          }
```

Comment:
Just tried it out, it's pretty neat. I don't know if I'd write `since="funky 0.5"`. But as a future feature, the compiler knows where classes come from, and sbt knows about versions. The tool can tell me `since 0.5 of funky, and you're using 1.0 already.` That would be handy when you're trying out recipes and you don't even know what dependency an import comes from. Then you'd realize you're reading an outdated blog post.

A trivial improvement would be to sort by `since` in the summary?